File: src/main/java/de/pflugradts/pwman3/application/commandhandling/handler/ViewCommandHandler.java
package de.pflugradts.pwman3.application.commandhandling.handler;

import com.google.common.eventbus.Subscribe;
import com.google.inject.Inject;
import de.pflugradts.pwman3.application.UserInterfaceAdapterPort;
import de.pflugradts.pwman3.application.commandhandling.command.ViewCommand;
import de.pflugradts.pwman3.domain.model.transfer.Output;
import de.pflugradts.pwman3.domain.service.PasswordService;

public class ViewCommandHandler implements CommandHandler {

    @Inject
    private PasswordService passwordService;
    @Inject
    private UserInterfaceAdapterPort userInterfaceAdapterPort;

    @Subscribe
    private void handleViewCommand(final ViewCommand viewCommand) {
        passwordService.viewPassword(viewCommand.getArgument()).ifPresent(passwordBytes -> {
            userInterfaceAdapterPort.send(Output.of(passwordBytes));
        });
        viewCommand.invalidateInput();
        userInterfaceAdapterPort.sendLineBreak();
    }

}

File: src/main/kotlin/de/pflugradts/passbird/application/commandhandling/handler/egg/ViewCommandHandler.kt
package de.pflugradts.passbird.application.commandhandling.handler.egg
import de.pflugradts.passbird.application.UserInterfaceAdapterPort
import de.pflugradts.passbird.application.commandhandling.CommandExecutionTracker
import de.pflugradts.passbird.application.commandhandling.command.ViewCommand
import de.pflugradts.passbird.application.commandhandling.handler.TypedCommandHandler
import de.pflugradts.passbird.application.useScrambled
import de.pflugradts.passbird.domain.model.transfer.Output.Companion.outputOf
import de.pflugradts.passbird.domain.service.password.PasswordService
class ViewCommandHandler constructor(
    private val passwordService: PasswordService,
    private val userInterfaceAdapterPort: UserInterfaceAdapterPort,
    private val commandExecutionTracker: CommandExecutionTracker,
) : TypedCommandHandler<ViewCommand>(ViewCommand::class.java) {
    override fun handleCommand(command: ViewCommand) {
        passwordService.viewPassword(command.argument).ifPresentOrElse(
            block = { it.useScrambled { shell -> userInterfaceAdapterPort.send(outputOf(shell)) } },
            other = { commandExecutionTracker.markFailure() },
        )
        command.invalidateInput()
        userInterfaceAdapterPort.sendLineBreak()
    }
}
